File: C++/037.cpp
#include "Utils.h"

using namespace std;

bool isLeftTruncatablePrime(string s) {
  int len = s.size();
  for (int trash=0; trash<len-1; trash++) {
    s = s.substr(1, s.size()-1); // remove left digit
    if (!isprime(stoi(s))) {
      return false;
    }
  }
  return true;
}

int main() {
  string primes[] = {"2", "3", "5", "7"};
  char add[] = {'1','3','7','9'};
  vector<string> products;
  for (auto i: primes) {
    for (auto j: add)
      products.push_back(i+j);
  }
  vector<string> candidates;

  for (int trash=0; trash<7; trash++) {
    for (int i=products.size()-1; i>=0; i--) {
      if (!isprime(stoi(products[i]))) {
        products.erase(products.begin()+i);
      }
    }
    for (auto i: products)
      cout << i << ", ";
    cout << "\n\n";

    candidates.insert(candidates.end(), products.begin(), products.end());
    vector<string> replacement;
    for (auto i: products) {
      for (auto j: add) {
        replacement.push_back(i+j);
      }
    }
    products = replacement;
  }

  int sum = 0;
  for (auto i: candidates) {
    if (isLeftTruncatablePrime(i)) {
      sum += stoi(i);
    }

  }
  cout <<  "\n\n\n" << sum;
}

File: Python/037.py
from itertools import product
from sympy import isprime

def remove_left_digit(n):
    return int(str(n)[1:])

def is_left_truncatable_prime(n):
    for _ in range(len(str(n))-1):
        n = remove_left_digit(n)
        if not isprime(n):
            return False
    return True

primes = ['2','3','5','7']
add = ['1','3','7','9']
products = [''.join(i) for i in product(primes,add)]
candidates = []

for _ in range(7):
    for j in reversed(products):
        if not isprime(j):
            products.remove(j)
    candidates += products
    products = [''.join(k) for k in product(products,add)]

print(sum(int(i) for i in candidates if is_left_truncatable_prime(int(i))))

for i in candidates:
    if is_left_truncatable_prime(int(i)):
        print(i)
